File: bench/cs/Ssh.Benchmark/SessionSetupBenchmark.cs

using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Security.Claims;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.Messages;
using Microsoft.DevTunnels.Ssh.Tcp;

namespace Microsoft.DevTunnels.Ssh.Benchmark;

class SessionSetupBenchmark : Benchmark
{
	private const string ConnectTimeMeasurement = "Connect time (ms)";
	private const string EncryptTimeMeasurement = "Encrypt time (ms)";
	private const string AuthTimeMeasurement = "Authenticate time (ms)";
	private const string ChannelTimeMeasurement = "Channnel open time (ms)";
	private const string TotalTimeMeasurement = "Total setup time (ms)";
	private const string BytesAllocatedMeasurement = "Bytes allocated (KB)";
	private const string BytesCopiedMeasurement = "Bytes copied (KB)";
	private const string LatencyMeasurement = "Latency (ms)";

	private static readonly TimeSpan latency = TimeSpan.FromMilliseconds(100);

	private readonly SshServer server;
	private readonly Task serverTask;
	private readonly SshClient client;

	public SessionSetupBenchmark(bool withLatency)
		: base("Session setup" + (withLatency ? " with latency" : ""))
	{
		HigherIsBetter[ConnectTimeMeasurement] = false;
		HigherIsBetter[EncryptTimeMeasurement] = false;
		HigherIsBetter[AuthTimeMeasurement] = false;
		HigherIsBetter[ChannelTimeMeasurement] = false;
		HigherIsBetter[TotalTimeMeasurement] = false;
		HigherIsBetter[BytesAllocatedMeasurement] = false;
		HigherIsBetter[BytesCopiedMeasurement] = false;
		HigherIsBetter[LatencyMeasurement] = false;

		var config = SshSessionConfiguration.DefaultWithReconnect;

		var trace = new TraceSource(nameof(SessionSetupBenchmark));
		this.server = withLatency
			? new SshServerWithLatency(config, trace)
			: new SshServer(config, trace);
		this.client = withLatency
			? new SshClientWithLatency(config, trace)
			: new SshClient(config, trace);

		var serverKey = SshAlgorithms.PublicKey.RsaWithSha512.GenerateKeyPair();
		this.server.Credentials = new[] { serverKey };

		this.server.SessionAuthenticating += OnServerSessionAuthenticating;
		this.server.ChannelRequest += OnServerChannelRequest;

		this.serverTask = this.server.AcceptSessionsAsync(Benchmark.ServerPort, IPAddress.Loopback);
	}

	protected override async Task RunAsync(Stopwatch stopwatch)
	{
#if DEBUG
		Buffer.Allocations.Clear();
		Buffer.Copies.Clear();
#endif

		double connectMark = 0;
		EventHandler<SshServerSession> sessionOpenedHandler = null;
		sessionOpenedHandler = (sender, serverSession) =>
		{
			connectMark = stopwatch.Elapsed.TotalMilliseconds;
			this.server.SessionOpened -= sessionOpenedHandler;
		};
		this.server.SessionOpened += sessionOpenedHandler;

		var clientSession = await client.OpenSessionAsync(
			IPAddress.Loopback.ToString(), ServerPort);

		var encryptMark = stopwatch.Elapsed.TotalMilliseconds;

		clientSession.Authenticating += OnClientSessionAuthenticating;
		await clientSession.AuthenticateServerAsync();

		var clientAuthCompletion = new TaskCompletionSource<bool>();
		await clientSession.AuthenticateClientAsync(
			("benchmark", "benchmark"), clientAuthCompletion);

		var authMark = stopwatch.Elapsed.TotalMilliseconds;

		var channelRequest = new ChannelRequestMessage
		{
			RequestType = "benchmark",
			WantReply = true,
		};

		// Protocol extension: Send initial request when opening channel.
		var clientChannel = await clientSession.OpenChannelAsync(
			new ChannelOpenMessage(), channelRequest);

		await clientAuthCompletion.Task;

		var channelMark = stopwatch.Elapsed.TotalMilliseconds;

		AddMeasurement(ConnectTimeMeasurement, connectMark);
		AddMeasurement(EncryptTimeMeasurement, encryptMark - connectMark);
		AddMeasurement(AuthTimeMeasurement, authMark - encryptMark);
		AddMeasurement(ChannelTimeMeasurement, channelMark - authMark);
		AddMeasurement(TotalTimeMeasurement, channelMark);

		// Add an additional request-reply to enable latency measurement,
		// which doesn't start until after the extension-info exchange.
		await clientChannel.RequestAsync(
			new ChannelRequestMessage { RequestType = "benchmark", WantReply = true, });

		AddMeasurement(LatencyMeasurement, clientSession.Metrics.LatencyAverageMs);

#if DEBUG
		lock (Buffer.Allocations)
		{
			long bytesAllocated = Buffer.Allocations.Values
				.Select((list) => list.Select(count => (long)count).Sum()).Sum();
			double kilobytesAllocated = (double)bytesAllocated / 1024;
			AddMeasurement(BytesAllocatedMeasurement, kilobytesAllocated);
		}

		lock (Buffer.Copies)
		{
			long bytesCopied = Buffer.Copies.Values
				.Select((list) => list.Select(count => (long)count).Sum()).Sum();
			double kilobytesCopied = (double)bytesCopied / 1024;
			AddMeasurement(BytesCopiedMeasurement, kilobytesCopied);
		}
#endif
	}

	private void OnClientSessionAuthenticating(object sender, SshAuthenticatingEventArgs e)
	{
		e.AuthenticationTask = Task.FromResult(new ClaimsPrincipal());
	}

	private void OnServerSessionAuthenticating(object sender, SshAuthenticatingEventArgs e)
	{
		e.AuthenticationTask = Task.FromResult(new ClaimsPrincipal());
	}

	private void OnServerChannelRequest(object sender, SshRequestEventArgs<ChannelRequestMessage> e)
	{
		e.IsAuthorized = true;
	}

	public override async ValueTask DisposeAsync()
	{
		this.server.Dispose();
		this.client.Dispose();
		await this.serverTask;
	}

	private class SshServerWithLatency : SshServer
	{
		public SshServerWithLatency(SshSessionConfiguration config, TraceSource trace) : base(config, trace) { }

		protected override async Task<(Stream Stream, IPAddress RemoteIPAddress)> AcceptConnectionAsync(TcpListener listener)
		{
			(var stream, var ipAddress) = await base.AcceptConnectionAsync(listener);
			return stream == null ? (null, null) : (new SlowStream(stream, SessionSetupBenchmark.latency), ipAddress);
		}
	}

	private class SshClientWithLatency : SshClient
	{
		public SshClientWithLatency(SshSessionConfiguration config, TraceSource trace) : base(config, trace) { }

		protected override async Task<(Stream Stream, IPAddress RemomoteIPAddress)> OpenConnectionAsync(
			string host, int port, CancellationToken cancellation)
		{
			(var stream, var ipAddress) = await base.OpenConnectionAsync(host, port, cancellation);
			return (new SlowStream(stream,SessionSetupBenchmark.latency), ipAddress);
		}
	}
}

File: bench/ts/ssh-bench/sessionSetupBenchmark.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import * as net from 'net';
import { Benchmark } from './benchmark';
import {
	SshClientCredentials,
	SshSessionConfiguration,
	SshAlgorithms,
	ChannelRequestMessage,
	Stream,
	SshProtocolExtensionNames,
} from '@microsoft/dev-tunnels-ssh';
import { SshClient, SshServer } from '@microsoft/dev-tunnels-ssh-tcp';
import { SlowStream } from './slowStream';
import { CancellationToken } from 'vscode-jsonrpc';

const ConnectTimeMeasurement = 'Connect time (ms)';
const EncryptTimeMeasurement = 'Encrypt time (ms)';
const AuthTimeMeasurement = 'Authenticate time (ms)';
const ChannelTimeMeasurement = 'Channnel open time (ms)';
const TotalTimeMeasurement = 'Total setup time (ms)';
const LatencyMeasurement = 'Latency (ms)';

declare type hrtime = [number, number];
const millis = ([s, ns]: hrtime) => s * 1000 + ns / 1000000;

const latencyInMilliseconds = 100;

export class SessionSetupBenchmark extends Benchmark {
	private readonly port: number;
	private readonly initPromise: Promise<void>;
	private readonly server: SshServer;
	private readonly serverPromise: Promise<void>;
	private readonly client: SshClient;

	public constructor(withLatency: boolean) {
		super('Session setup' + (withLatency ? ' with latency' : ''));

		this.higherIsBetter.set(ConnectTimeMeasurement, false);
		this.higherIsBetter.set(EncryptTimeMeasurement, false);
		this.higherIsBetter.set(AuthTimeMeasurement, false);
		this.higherIsBetter.set(ChannelTimeMeasurement, false);
		this.higherIsBetter.set(TotalTimeMeasurement, false);
		this.higherIsBetter.set(LatencyMeasurement, false);

		const config = new SshSessionConfiguration();
		config.protocolExtensions.push(SshProtocolExtensionNames.sessionReconnect);
		config.protocolExtensions.push(SshProtocolExtensionNames.sessionLatency);

		this.server = withLatency ? new SshServerWithLatency(config) : new SshServer(config);
		this.client = withLatency ? new SshClientWithLatency(config) : new SshClient(config);

		this.initPromise = SshAlgorithms.publicKey
			.rsaWithSha512!.generateKeyPair()
			.then((serverKey) => {
				this.server.credentials.publicKeys.push(serverKey);
			});

		this.server.onSessionOpened((session) => {
			session.onAuthenticating((e) => {
				e.authenticationPromise = Promise.resolve({});
			});
			session.onChannelOpening((e) => {
				e.channel.onRequest((e) => {
					e.isAuthorized = true;
				});
			});
		});

		this.port = Benchmark.findAvailablePort();
		this.serverPromise = this.server.acceptSessions(this.port, 'localhost');
	}

	public async run(): Promise<void> {
		await this.initPromise;

		const startTime: hrtime = process.hrtime();

		let connectMark: hrtime = [0, 0];
		const sessionOpenedRegistration = this.server.onSessionOpened((session) => {
			connectMark = process.hrtime(startTime);
			sessionOpenedRegistration.dispose();
		});

		var clientSession = await this.client.openSession('localhost', this.port);

		var encryptMark: hrtime = process.hrtime(startTime);

		clientSession.onAuthenticating((e) => {
			e.authenticationPromise = Promise.resolve({});
		});

		await clientSession.authenticateServer();

		let clientAuthCallback: (err?: Error, result?: boolean) => void;
		const clientAuthPromise = new Promise((resolve, reject) => {
			clientAuthCallback = (err, result) => {
				if (err) reject(err);
				else resolve(result);
			};
		});

		const credentials: SshClientCredentials = { username: 'benchmark', password: 'benchmark' };
		await clientSession.authenticateClient(credentials, clientAuthCallback!);

		var authMark: hrtime = process.hrtime(startTime);

		var channelRequest = new ChannelRequestMessage('benchmark');
		channelRequest.wantReply = true;

		// Protocol extension: Send initial request when opening channel.
		const clientChannel = await clientSession.openChannel(null, channelRequest);

		await clientAuthPromise;

		var channelMark: hrtime = process.hrtime(startTime);

		this.addMeasurement(ConnectTimeMeasurement, millis(connectMark));
		this.addMeasurement(EncryptTimeMeasurement, millis(encryptMark) - millis(connectMark));
		this.addMeasurement(AuthTimeMeasurement, millis(authMark) - millis(encryptMark));
		this.addMeasurement(ChannelTimeMeasurement, millis(channelMark) - millis(authMark));
		this.addMeasurement(TotalTimeMeasurement, millis(channelMark));

		// Add an additional request-reply to enable latency measurement,
		// which doesn't start until after the extension-info exchange.
		const channelRequest2 = new ChannelRequestMessage();
		channelRequest2.requestType = 'test';
		channelRequest2.wantReply = true;
		await clientChannel.request(channelRequest2);

		this.addMeasurement(LatencyMeasurement, clientSession.metrics.latencyAverageMs);
	}

	public async dispose(): Promise<void> {
		this.server.dispose();
		this.client.dispose();
		await this.serverPromise;
	}
}

class SshServerWithLatency extends SshServer {
	public constructor(config: SshSessionConfiguration) {
		super(config);
	}

	protected async acceptConnection(socket: net.Socket): Promise<Stream> {
		const baseStream = await super.acceptConnection(socket);
		return new SlowStream(baseStream, latencyInMilliseconds);
	}
}

class SshClientWithLatency extends SshClient {
	public constructor(config: SshSessionConfiguration) {
		super(config);
	}

	protected async openConnection(
		serverHost: string,
		serverPort?: number,
		cancellation?: CancellationToken,
	): Promise<{ stream: Stream; ipAddress: string | undefined }> {
		const connectionResult = await super.openConnection(serverHost, serverPort, cancellation);
		return {
			stream: new SlowStream(connectionResult.stream, latencyInMilliseconds),
			ipAddress: connectionResult.ipAddress
		};
	}
}
